File: implementations/erlang/apps/ff3_validate/src/ff3_validate.erl
%% FF3 NIST Test Vector Validator Main Module

-module(ff3_validate).
-export([main/1]).

-record(opts, {
    vectors_path = undefined :: undefined | string(),
    verbose = false :: boolean(),
    quiet = false :: boolean(),
    help = false :: boolean()
}).

-record(nist_vector, {
    sample :: integer(),
    algorithm :: string(),
    key :: string(),
    radix :: integer(),
    alphabet :: string(),
    plaintext :: string(),
    tweak :: string(),
    ciphertext :: string()
}).

main(Args) ->
    try
        Opts = parse_args(Args, #opts{}),

        case Opts#opts.help of
            true ->
                show_usage(),
                halt(0);
            false ->
                run_validation(Opts)
        end
    catch
        error:{badarg, Msg} ->
            io:format(standard_error, "Error: ~s~n", [Msg]),
            show_usage(),
            halt(1);
        Error:Reason ->
            io:format(standard_error, "Error: ~p:~p~n", [Error, Reason]),
            halt(2)
    end.

show_usage() ->
    io:format("FF3 NIST Test Vector Validation Tool~n~n"),
    io:format("Usage: ff3-validate [OPTIONS]~n~n"),
    io:format("Options:~n"),
    io:format("  --vectors PATH    Path to test vectors JSON file~n"),
    io:format("  --verbose         Show detailed test output~n"),
    io:format("  --quiet           Only show failures and summary~n"),
    io:format("  -h, --help        Show this help message~n~n").

parse_args([], Opts) -> Opts;
parse_args(["-h"|_], Opts) -> Opts#opts{help = true};
parse_args(["--help"|_], Opts) -> Opts#opts{help = true};
parse_args(["--vectors", Path|Rest], Opts) ->
    parse_args(Rest, Opts#opts{vectors_path = Path});
parse_args(["--verbose"|Rest], Opts) ->
    parse_args(Rest, Opts#opts{verbose = true});
parse_args(["--quiet"|Rest], Opts) ->
    parse_args(Rest, Opts#opts{quiet = true});
parse_args([Arg|_], _) ->
    error({badarg, io_lib:format("Unknown option: ~s", [Arg])}).

find_vectors_file(undefined) ->
    Paths = ["../../shared/test-vectors/nist_ff3_official_vectors.json",
             "../../../shared/test-vectors/nist_ff3_official_vectors.json",
             "./nist_ff3_official_vectors.json"],
    find_first_existing(Paths);
find_vectors_file(Path) ->
    case filelib:is_file(Path) of
        true -> {ok, Path};
        false -> {error, "Vectors file not found: " ++ Path}
    end.

find_first_existing([]) -> {error, "Could not find NIST test vectors file"};
find_first_existing([Path|Rest]) ->
    case filelib:is_file(Path) of
        true -> {ok, Path};
        false -> find_first_existing(Rest)
    end.

run_validation(Opts) ->
    Quiet = Opts#opts.quiet,

    case not Quiet of
        true ->
            io:format("FF3 NIST Test Vector Validation Tool~n"),
            io:format("========================================~n~n");
        false -> ok
    end,

    case find_vectors_file(Opts#opts.vectors_path) of
        {error, Msg} ->
            io:format(standard_error, "Error: ~s~n", [Msg]),
            case Opts#opts.vectors_path of
                undefined ->
                    io:format(standard_error, "Try: ff3-validate --vectors /path/to/vectors.json~n", []);
                _ -> ok
            end,
            halt(2);
        {ok, VectorsPath} ->
            case not Quiet of
                true -> io:format("Vector file: ~s~n~n", [VectorsPath]);
                false -> ok
            end,

            Vectors = load_nist_vectors(VectorsPath),

            case not Quiet of
                true -> io:format("Testing ~p NIST FF3 vectors...~n~n", [length(Vectors)]);
                false -> ok
            end,

            {Passed, Total} = run_tests(Vectors, Opts, 0, 0),

            case not Quiet of
                true ->
                    io:format("~n========================================~n"),
                    io:format("Results: ~p/~p passed~n~n", [Passed, Total]);
                false -> ok
            end,

            case Passed =:= Total of
                true ->
                    case not Quiet of
                        true ->
                            io:format("ALL NIST TEST VECTORS PASSED!~n~n"),
                            io:format("WARNING: FF3 was WITHDRAWN by NIST due to security vulnerabilities.~n"),
                            io:format("This implementation is for EDUCATIONAL and RESEARCH purposes only.~n"),
                            io:format("DO NOT use in production systems.~n~n");
                        false -> ok
                    end,
                    halt(0);
                false ->
                    case not Quiet of
                        true -> io:format("VALIDATION FAILED~n~n");
                        false -> ok
                    end,
                    halt(1)
            end
    end.

run_tests([], _Opts, Passed, Total) -> {Passed, Total};
run_tests([Vector | Rest], Opts, Passed, Total) ->
    case test_single_vector(Vector, Opts) of
        pass -> run_tests(Rest, Opts, Passed + 1, Total + 1);
        fail -> run_tests(Rest, Opts, Passed, Total + 1)
    end.

test_single_vector(Vector, Opts) ->
    #nist_vector{
        sample = Sample,
        key = KeyHex,
        radix = Radix,
        plaintext = Plaintext,
        tweak = TweakHex,
        ciphertext = ExpectedCiphertext
    } = Vector,

    try
        Key = ff3_api:hex_to_bytes(KeyHex),
        Tweak = ff3_api:hex_to_bytes(TweakHex),

        Alphabet = case Radix of
            10 -> ff3_alphabets:digits_alphabet();
            26 -> ff3_alphabets:custom_radix26_alphabet();
            _ -> error({unsupported_radix, Radix})
        end,

        Result = ff3_api:encrypt(Plaintext, Key, Tweak, Alphabet),
        Decrypted = ff3_api:decrypt(Result, Key, Tweak, Alphabet),

        EncryptPassed = (Result =:= ExpectedCiphertext),
        RoundtripPassed = (Decrypted =:= Plaintext),
        TestPassed = EncryptPassed andalso RoundtripPassed,

        case TestPassed of
            true ->
                case Opts#opts.verbose of
                    true -> io:format("Sample ~p: PASS~n", [Sample]);
                    false -> ok
                end,
                pass;
            false ->
                case Opts#opts.quiet of
                    false ->
                        io:format("Sample ~p: FAIL~n", [Sample]),
                        case EncryptPassed of
                            false ->
                                io:format("  Expected: ~s~n", [ExpectedCiphertext]),
                                io:format("  Got:      ~s~n", [Result]);
                            true -> ok
                        end,
                        case RoundtripPassed of
                            false -> io:format("  Round-trip failed~n", []);
                            true -> ok
                        end;
                    true -> ok
                end,
                fail
        end
    catch
        _:Error ->
            case Opts#opts.quiet of
                false -> io:format("Sample ~p: ERROR - ~p~n", [Sample, Error]);
                true -> ok
            end,
            fail
    end.

load_nist_vectors(FilePath) ->
    {ok, Binary} = file:read_file(FilePath),
    Json = jsx:decode(Binary, [return_maps]),
    #{<<"vectors">> := VectorsList} = Json,
    [parse_vector(V) || V <- VectorsList].

parse_vector(VectorMap) ->
    #nist_vector{
        sample = maps:get(<<"sample">>, VectorMap),
        algorithm = binary_to_list(maps:get(<<"algorithm">>, VectorMap)),
        key = binary_to_list(maps:get(<<"key">>, VectorMap)),
        radix = maps:get(<<"radix">>, VectorMap),
        alphabet = binary_to_list(maps:get(<<"alphabet">>, VectorMap, <<"">>)),
        plaintext = binary_to_list(maps:get(<<"plaintext">>, VectorMap)),
        tweak = binary_to_list(maps:get(<<"tweak">>, VectorMap)),
        ciphertext = binary_to_list(maps:get(<<"ciphertext">>, VectorMap))
    }.

File: implementations/cpp/apps/ff3_validate.cpp
/**
 * @file ff3_validate.cpp
 * @brief FF3 NIST Test Vector Validation Tool
 */

#include <ff3.hpp>
#include <iostream>
#include <fstream>
#include <cstring>
#include <filesystem>

using namespace ff3;

struct Options {
    std::string vectors_path;
    bool verbose = false;
    bool quiet = false;
    bool help = false;
};

struct TestVector {
    int sample;
    std::string algorithm;
    std::string key;
    int radix;
    std::string plaintext;
    std::string tweak;
    std::string ciphertext;
};

void show_usage() {
    std::cout << "FF3 NIST Test Vector Validation Tool\n\n";
    std::cout << "Usage: ff3-validate [OPTIONS]\n\n";
    std::cout << "Options:\n";
    std::cout << "  --vectors PATH    Path to test vectors JSON file\n";
    std::cout << "  --verbose         Show detailed test output\n";
    std::cout << "  --quiet           Only show failures and summary\n";
    std::cout << "  -h, --help        Show this help message\n\n";
}

Options parse_args(int argc, char* argv[]) {
    Options opts;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
            opts.help = true;
            return opts;
        } else if (strcmp(argv[i], "--vectors") == 0) {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --vectors");
            }
            opts.vectors_path = argv[++i];
        } else if (strcmp(argv[i], "--verbose") == 0) {
            opts.verbose = true;
        } else if (strcmp(argv[i], "--quiet") == 0) {
            opts.quiet = true;
        } else {
            throw std::runtime_error(std::string("Unknown option: ") + argv[i]);
        }
    }

    return opts;
}

std::string find_vectors_file(const std::string& custom_path) {
    if (!custom_path.empty()) {
        if (std::filesystem::exists(custom_path)) {
            return custom_path;
        }
        throw std::runtime_error("Vectors file not found: " + custom_path);
    }

    std::vector<std::string> paths = {
        "../../shared/test-vectors/nist_ff3_official_vectors.json",
        "../../../shared/test-vectors/nist_ff3_official_vectors.json",
        "./nist_ff3_official_vectors.json"
    };

    for (const auto& path : paths) {
        if (std::filesystem::exists(path)) {
            return path;
        }
    }

    throw std::runtime_error("Could not find NIST test vectors file");
}

std::vector<TestVector> load_vectors(const std::string& filepath) {
    std::ifstream file(filepath);
    if (!file.is_open()) {
        throw std::runtime_error("Cannot open vectors file: " + filepath);
    }

    std::vector<TestVector> vectors;
    std::string line;
    TestVector current;

    while (std::getline(file, line)) {
        if (line.find("\"sample\":") != std::string::npos) {
            size_t pos = line.find(":") + 1;
            current.sample = std::stoi(line.substr(pos));
        } else if (line.find("\"key\":") != std::string::npos) {
            size_t start = line.find("\"", line.find(":")) + 1;
            size_t end = line.find("\"", start);
            current.key = line.substr(start, end - start);
        } else if (line.find("\"radix\":") != std::string::npos) {
            size_t pos = line.find(":") + 1;
            current.radix = std::stoi(line.substr(pos));
        } else if (line.find("\"plaintext\":") != std::string::npos) {
            size_t start = line.find("\"", line.find(":")) + 1;
            size_t end = line.find("\"", start);
            current.plaintext = line.substr(start, end - start);
        } else if (line.find("\"tweak\":") != std::string::npos) {
            size_t start = line.find("\"", line.find(":")) + 1;
            size_t end = line.find("\"", start);
            current.tweak = line.substr(start, end - start);
        } else if (line.find("\"ciphertext\":") != std::string::npos) {
            size_t start = line.find("\"", line.find(":")) + 1;
            size_t end = line.find("\"", start);
            current.ciphertext = line.substr(start, end - start);
            vectors.push_back(current);
        }
    }

    return vectors;
}

int main(int argc, char* argv[]) {
    try {
        Options opts = parse_args(argc, argv);

        if (opts.help) {
            show_usage();
            return 0;
        }

        if (!opts.quiet) {
            std::cout << "FF3 NIST Test Vector Validation Tool\n";
            std::cout << "========================================\n\n";
        }

        std::string vectors_path = find_vectors_file(opts.vectors_path);

        if (!opts.quiet) {
            std::cout << "Vector file: " << vectors_path << "\n\n";
        }

        auto vectors = load_vectors(vectors_path);

        if (!opts.quiet) {
            std::cout << "Testing " << vectors.size() << " NIST FF3 vectors...\n\n";
        }

        int passed = 0;
        int total = 0;

        for (const auto& vec : vectors) {
            // Only test radix 10 and 26 (skip unsupported radixes)
            if (vec.radix != 10 && vec.radix != 26) {
                continue;
            }

            total++;

            try {
                const auto key = hex_to_bytes(vec.key);
                const auto tweak = hex_to_bytes(vec.tweak);

                FF3 cipher = (vec.radix == 10) ?
                    digits(key, tweak) :
                    radix26(key, tweak);

                const auto encrypted = cipher.encrypt(vec.plaintext);
                const auto decrypted = cipher.decrypt(encrypted);

                bool encrypt_passed = (encrypted == vec.ciphertext);
                bool roundtrip_passed = (decrypted == vec.plaintext);
                bool test_passed = encrypt_passed && roundtrip_passed;

                if (test_passed) {
                    passed++;
                    if (opts.verbose) {
                        std::cout << "Sample " << vec.sample << ": PASS\n";
                    }
                } else {
                    if (!opts.quiet) {
                        std::cout << "Sample " << vec.sample << ": FAIL\n";
                        if (!encrypt_passed) {
                            std::cout << "  Expected: " << vec.ciphertext << "\n";
                            std::cout << "  Got:      " << encrypted << "\n";
                        }
                        if (!roundtrip_passed) {
                            std::cout << "  Round-trip failed\n";
                        }
                    }
                }
            } catch (const std::exception& e) {
                if (!opts.quiet) {
                    std::cout << "Sample " << vec.sample << ": ERROR - " << e.what() << "\n";
                }
            }
        }

        if (!opts.quiet) {
            std::cout << "\n========================================\n";
            std::cout << "Results: " << passed << "/" << total << " passed\n";

            if (passed == total) {
                std::cout << "\nALL NIST TEST VECTORS PASSED!\n\n";
                std::cout << "WARNING: FF3 was WITHDRAWN by NIST due to security vulnerabilities.\n";
                std::cout << "This implementation is for EDUCATIONAL and RESEARCH purposes only.\n";
                std::cout << "DO NOT use in production systems.\n\n";
            } else {
                std::cout << "\nVALIDATION FAILED\n";
            }
        }

        return (passed == total) ? 0 : 1;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 2;
    }
}
